### PR TITLE
Feature/save transfer draft

### PR DESCRIPTION
**Pull request by JohnPadgettCivica**

### CODE REVIEW

File: src/EA.Weee.RequestHandlers/AatfEvidence/TransferTonnagesValidator.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+        public async Task Validate(List<TransferTonnageValue> transferValues)
+        {
+            var existingNoteTonnage =
+                await evidenceDataAccess.GetTonnageByIds(transferValues.Select(t => t.TransferTonnageId)
+                    .ToList());
+
+            foreach (var noteTonnage in existingNoteTonnage)
+            {
+                // can't
```
Comment (JohnPadgettCivica) [inline src/EA.Weee.RequestHandlers/AatfEvidence/TransferTonnagesValidator.cs:28]: This is removed in subsequent change

File: src/EA.Weee.RequestHandlers.Tests.Unit/AatfEvidence/TransferTonnagesValidatorTests.cs [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+            //act
+            var exception = await Record.ExceptionAsync(async () => await validator.Validate(transferValues));
+
+            //assert
+            exception.Should().BeOfType<InvalidOperationException>();
+        }
+
+        [Theory]
+        [ClassData(typeof(NoteStatusData))]
+        public async Task Validate_GivenRequestedTonnageIsAvailableAsNoteIsNoteTransferIsNotApproved_NoExceptionExpected(NoteStatus status)
```
Comment (meglenapnovoselska) [inline src/EA.Weee.RequestHandlers.Tests.Unit/AatfEvidence/TransferTonnagesValidatorTests.cs:174]: would you mind making it a bit more readable?  the middle bit

File: src/EA.Weee.Requests/Scheme/TransferEvidenceNoteRequest.cs [hunk 3]
```diff
@@ -43,6 +43,20 @@ public TransferEvidenceNoteRequest()
             EvidenceNoteIds = evidenceNoteIds;
         }
 
+        public TransferEvidenceNoteRequest(Guid organisationId,
+            Guid schemeId,
+            List<TransferTonnageValue> transferValues,
```
Comment (meglenapnovoselska) [inline src/EA.Weee.Requests/Scheme/TransferEvidenceNoteRequest.cs:48]: no Condition for them, they could be null?!

File: src/EA.Weee.Integration.Tests/Handlers/CreateTransferEvidenceNoteHandlerIntegrationTests.cs [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+                };
+
+                TransferEvidenceNoteDbSetup.Init().With(t =>
+                {
+                    t.UpdateStatus(NoteStatus.Submitted, UserId.ToString());
+                }).WithTonnages(newTransferNoteTonnage3).Create();
+
+                // now to call the request to take the last 1 tonne
+                transferTonnageValues = new List<TransferTonnageValue>()
+                {
```
[the remark this answers is not in the document]
Comment (JohnPadgettCivica) [reply src/EA.Weee.Integration.Tests/Handlers/CreateTransferEvidenceNoteHandlerIntegrationTests.cs:152]: The newTransferNoteTonnage2 list gets added directly to the database in the following bit of code,
TransferEvidenceNoteDbSetup.Init().With(t =>
                {
                    t.UpdateStatus(NoteStatus.Submitted, UserId.ToString());
                    t.UpdateStatus(NoteStatus.Approved, UserId.ToString());
                }).WithTonnages(newTransferNoteTonnage2).Create();
It's just to act as if another user has created a note and stolen the tonnage while someone else is using the system. Probably very unlikely to happen in real use.

File: src/EA.Weee.RequestHandlers.Tests.DataAccess/Evidence/EvidenceDataAccessUnitTests.cs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+                    c.Received.Equals(requestTonnageValue.Received)) != null))).MustHaveHappenedOnceExactly();
+            }
+
+            SystemTime.Unfreeze();
+        }
+
+        [Theory]
+        [InlineData(Core.AatfEvidence.NoteStatus.Submitted)]
+        [InlineData(Core.AatfEvidence.NoteStatus.Draft)]
+        public async Task HandleAsync_GivenTransferNote_NoteShouldBeAddedAndSaveChangesCalled(Core.AatfEvidence.NoteStatus status)
```
[the remark this answers is not in the document]
Comment (JohnPadgettCivica) [reply src/EA.Weee.RequestHandlers.Tests.DataAccess/Evidence/EvidenceDataAccessUnitTests.cs:134]: Yeah agree. Is this in reference to a particular test here? Some of them have quite complex asserts so have split them out.

[the remark this answers is not in the document]
Comment (meglenapnovoselska) [reply src/EA.Weee.RequestHandlers.Tests.DataAccess/Evidence/EvidenceDataAccessUnitTests.cs:134]: it's up to you. wherever is possible of course.

File: src/EA.Weee.RequestHandlers.Tests.Unit/AatfEvidence/CreateTransferEvidenceNoteRequestHandlerTests.cs [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+        {
+            //arrange
+            A.CallTo(() => genericDataAccess.GetById<Organisation>(A<Guid>._)).Returns((Organisation)null);
+
+            //act
+            var result = await Record.ExceptionAsync(() => handler.HandleAsync(Request()));
+
+            //assert
+            result.Should().BeOfType<ArgumentNullException>();
+        }
```
[the remark this answers is not in the document]
Comment (JohnPadgettCivica) [reply src/EA.Weee.RequestHandlers.Tests.Unit/AatfEvidence/CreateTransferEvidenceNoteRequestHandlerTests.cs:169]: In the next PR I have actually moved the exception into the  getById so will be removing many of these tests :)

File: src/EA.Weee.Requests/Scheme/TransferEvidenceNoteRequest.cs (hunk 3 above)
Comment (JohnPadgettCivica) [reply src/EA.Weee.Requests/Scheme/TransferEvidenceNoteRequest.cs:48]: done... will add unit tests for this in another PR

File: src/EA.Weee.RequestHandlers.Tests.Unit/AatfEvidence/TransferTonnagesValidatorTests.cs (hunk 2 above)
Comment (JohnPadgettCivica) [reply src/EA.Weee.RequestHandlers.Tests.Unit/AatfEvidence/TransferTonnagesValidatorTests.cs:180]: Tried to clean it up a little

File: src/EA.Weee.RequestHandlers.Tests.DataAccess/Evidence/EvidenceDataAccessUnitTests.cs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+                    c.NoteTonnageId.Equals(requestTonnageValue.NoteTonnageId) &&
+                    c.Reused.Equals(requestTonnageValue.Reused) &&
+                    c.Received.Equals(requestTonnageValue.Received)) != null))).MustHaveHappenedOnceExactly();
+            }
+
+            SystemTime.Unfreeze();
+        }
+
+        [Fact]
+        public async Task HandleAsync_GivenSubmittedTransferNote_NoteShouldBeAddedToContext()
```
[the remark this answers is not in the document]
Comment (JohnPadgettCivica) [reply src/EA.Weee.RequestHandlers.Tests.DataAccess/Evidence/EvidenceDataAccessUnitTests.cs]: done

File: src/EA.Weee.RequestHandlers.Tests.DataAccess/Evidence/EvidenceDataAccessUnitTests.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (JohnPadgettCivica) [reply src/EA.Weee.RequestHandlers.Tests.DataAccess/Evidence/EvidenceDataAccessUnitTests.cs]: done

File: src/EA.Weee.DataAccess.Tests.Integration/EvidenceNote/NotesDataAccessGetTonnageByIdsIntegration.cs [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    using Weee.DataAccess.DataAccess;
+    using Weee.Tests.Core;
+    using Weee.Tests.Core.Model;
+    using Xunit;
+
+    public class NotesDataAccessGetTonnageByIdsIntegration : EvidenceNoteBaseDataAccess
+    {
+        [Fact]
+        public async Task GetTonnageByIds_GivenIds_ShouldReturnCorrectNotes()
+        {
```
[the remark this answers is not in the document]
Comment (JohnPadgettCivica) [reply src/EA.Weee.DataAccess.Tests.Integration/EvidenceNote/NotesDataAccessGetTonnageByIdsIntegration.cs:23]: Added....

File: src/EA.Weee.DataAccess.Tests.Integration/EvidenceNote/NotesDataAccessGetTonnageByIdsIntegration.cs [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+                var noteTonnage = noteTonnages.First(nt => nt.Id.Equals(note1.NoteTonnage.ElementAt(1).Id));
+                noteTonnage.Should().NotBeNull();
+                noteTonnage.CategoryId.Should().Be(WeeeCategory.GasDischargeLampsAndLedLightSources);
+                noteTonnage.Received.Should().Be(2);
+                noteTonnage.Reused.Should().Be(1);
+                noteTonnage.Note.Should().NotBeNull();
+                noteTonnage.NoteTransferTonnage.Should().NotBeNull();
+                noteTonnage.NoteTransferTonnage.Select(nt => nt.TransferNote).Should().NotBeNull();
+
+                noteTonnage = noteTonnages.First(nt => nt.Id.Equals(note2.NoteTonnage.ElementAt(0).Id));
```
[the remark this answers is not in the document]
Comment (JohnPadgettCivica) [reply src/EA.Weee.DataAccess.Tests.Integration/EvidenceNote/NotesDataAccessGetTonnageByIdsIntegration.cs]: updated so in separate variables

**Review by meglenapnovoselska — APPROVED** [review]

**Pull request merged**